test/adaptive_load/plugin_util_test.cc

```diff
@@ -0,0 +1,378 @@
+#include "absl/status/status.h"
+#include "envoy/config/core/v3/base.pb.h"
+#include "envoy/registry/registry.h"
+
+#include "nighthawk/adaptive_load/input_variable_setter.h"
+#include "nighthawk/adaptive_load/metrics_plugin.h"
+#include "nighthawk/adaptive_load/scoring_function.h"
+#include "nighthawk/adaptive_load/step_controller.h"
+
+#include "external/envoy/source/common/config/utility.h"
+
+#include "api/adaptive_load/benchmark_result.pb.h"
+#include "api/adaptive_load/scoring_function_impl.pb.h"
+#include "api/client/options.pb.h"
+
+#include "adaptive_load/plugin_util.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace Nighthawk {
+
+namespace {
+
+const double kBadConfigThreshold = 98765.0;
+
+/**
+ * Returns a validation error if the config proto contains kBadConfigThreshold.
+ *
+ * @param message An Any proto that must wrap a LinearScoringFunctionConfig.
+ *
+ * @return Status InvalidArgument if threshold is kBadConfigThreshold, OK otherwise.
+ */
+absl::Status DoValidateConfig(const Envoy::Protobuf::Message& message) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  nighthawk::adaptive_load::LinearScoringFunctionConfig config;
+  Envoy::MessageUtil::unpackTo(any, config);
+  return config.threshold() == kBadConfigThreshold
+             ? absl::InvalidArgumentError("input validation failed")
+             : absl::OkStatus();
+}
+
+/**
+ * InputVariableSetter for testing.
+ */
+class TestInputVariableSetter : public InputVariableSetter {
+public:
+  // Any plugin in the adaptive load system can freely choose an arbitrary single proto as its
+  // config type. We use LinearScoringFunctionConfig for all plugins in this test.
+  TestInputVariableSetter(const nighthawk::adaptive_load::LinearScoringFunctionConfig& config)
+      : value_from_config_proto_{config.threshold()} {}
+  absl::Status SetInputVariable(nighthawk::client::CommandLineOptions& command_line_options,
+                                double input_value) override {
+    command_line_options.mutable_connections()->set_value(static_cast<unsigned int>(input_value));
+    return absl::OkStatus();
+  }
+  const double value_from_config_proto_;
+};
+
+/**
+ * A factory that creates a TestInputVariableSetter from a LinearScoringFunctionConfig (see
+ * TestInputVariableSetter constructor).
+ */
+class TestInputVariableSetterConfigFactory : public InputVariableSetterConfigFactory {
```

Comment:
Is this going to be the only unit test where we will need these "test only" elements? Or should we add them into their own testonly library so they are reusable?

Or is this something we want to address separately in #448 ?

Reply:
I think all of these can be replaced by the fakes created in #448.